File: dama-android/app/src/main/java/com/villager/mezi/dama/Player.java
package com.villager.mezi.dama;

import java.util.ArrayList;
import java.util.Arrays;

public class Player {

    public static final int[][] NEIGHBOUR_OFFSETS = {{-1, -1}, {-1, 1}, {1, 1}, {1, -1}};

    protected Board board;
    private boolean isTop;  // player's position
    private String name;
    private static boolean isEgregna = true;
    private static boolean allowEfita = true;
    private int[] lastMove;
    private ArrayList<int[]> couldHaveEatens;

    public Player(Board board, boolean isTop, String name){
        this.board = board;
        this.isTop = isTop;
        setName(name);
    }

    public Player(Board board, boolean isTop){
        this(board, isTop, isTop ? "Top Player" : "Bot Player");
    }

    public void setName(String name){
        this.name = name;
    }

    public String getName(){
        return name;
    }

    public static void setIsEgregna(boolean isItEgregna){
        Player.isEgregna = isItEgregna;
    }

    public static boolean isEgregna(){return isEgregna; }


    public static boolean isEfitaAllowed(){
        return allowEfita;
    }

    // calculates the number of corkis that a player has left
    // and returns  the number of corkis that a player has left on the board
    public int getNumCorkis() {
        int numCorkis = 0;
        for (int row = 0; row < Board.NUM_ROWS; row++) {
            for (int col = 0; col < Board.NUM_ROWS; col++ ) {
                if (isTop) {
                    numCorkis += board.isTop(row, col) ? 1 : 0;
                } else {
                    numCorkis += board.isBottom(row, col) ? 1 : 0;
                }
            }
        }
        return numCorkis;
    }

    public boolean isTop() {
        return  isTop;
    }

    // identifies the legal cells(eats and move destinations) that can be reached out from a given
    // corki and returns them wrapper in an array list of arrays, the array at index 0 containing
    // all the moves and the array at index 1 containing all the eats
    public ArrayList<int[]>[] movesAndEats(int row, int col)  {
        int[] curCell = {row, col};
        ArrayList<int[]>[] movesAndEats = new ArrayList[]{new ArrayList<>(), new ArrayList<>()};
        for(int[] offset: NEIGHBOUR_OFFSETS){
            int[] neighborCell = {row + offset[0], col + offset[1]};
            boolean isTankegnaKing = (!Player.isEgregna) && board.isKing(curCell);
            int[] prevNeighborCell = curCell;
            while(true) {
                if (board.isInsideBoard(neighborCell)) {
                    if (canMove(curCell, neighborCell)) {
                        movesAndEats[0].add(neighborCell);
                    } else {
                        int[] landingCell = isTankegnaKing ? board.getNextCell(prevNeighborCell,
                                neighborCell) : board.getNextCell(curCell, neighborCell);
                        if(landingCell != null) {
                            if (canEat(curCell, landingCell)) {
                                movesAndEats[1].add(landingCell);
                            }
                        }
                    }
                }
                if(!isTankegnaKing){
                    break;
                } else {
                    int temp[] = neighborCell.clone();
                    neighborCell = board.getNextCell(prevNeighborCell, neighborCell);
                    prevNeighborCell = temp;
                    if(neighborCell == null) break;
                }
            }
        }
        return movesAndEats;
    }

    public ArrayList<int[]>[] movesAndEats(int[] cell)  {
        return movesAndEats(cell[0], cell[1]);
    }

    public ArrayList<int[]> advancesFromCell(int[] fromCell)  {
        ArrayList<int[]>[] advances= movesAndEats(fromCell[0], fromCell[1]);
        if((!allowEfita) && advances[1].size() > 0) return advances[1];
        ArrayList<int[]> combined = new ArrayList<>();
        for (int[] cell: advances[0]) {
            combined.add(cell);
        }
        for (int[] cell: advances[1]) {
            combined.add(cell);
        }
        return combined;
    }


    public ArrayList<int[]> getCanAdvanceCells() {
        ArrayList<int[]> canAdvanceCells = new ArrayList<>();
        ArrayList<int[]> canEatCells = new ArrayList<>();

        for (int row = 0; row < Board.NUM_ROWS; row++) {
            for (int col = 0; col < Board.NUM_ROWS; col++) {
                if ((isTop && board.isTop(row, col) || ((!isTop) && board.isBottom(row, col)))) {
                    ArrayList[] movesAndEats = movesAndEats(row, col);
                    int numMoves = movesAndEats[0].size();
                    int numEats = movesAndEats[1].size();
                    if (numMoves > 0 || numEats > 0) {
                        int[] cell = {row, col};
                        canAdvanceCells.add(cell);
                        if (numEats > 0) {
                            canEatCells.add(cell);
                        }
                    }
                }
            }
        }
        if(!allowEfita) {
            if(canEatCells.size() > 0) return canEatCells;
            else return canAdvanceCells;
        }
        else {
            couldHaveEatens = canEatCells;
            return canAdvanceCells;
        }
    }

    public int[] getEfita(){
        if(couldHaveEatens.size() > 0){
            int[] potentialEfita = couldHaveEatens.get(0);
            if(board.isEmptyLegalCell(potentialEfita)) return lastMove;
            else return potentialEfita;
        }
        else return null;
    }

    public void eatEfita(int[] efita){
        board.makeEmpty(efita);
    }

    public boolean[] playTurn(int[] startCell, int[] landingCell, int[] eatenCorki){
        boolean didEat = false, wasKingified = false;
        if(canEat(startCell, landingCell)){
            didEat = true;
            int[] target = (!Player.isEgregna) && board.isKing(startCell) ?
                    board.getTankegnaKingsPrey(startCell, landingCell) :
                    board.getMiddleCell(startCell, landingCell);
            eatenCorki[0] = target[0]; eatenCorki[1] = target[1]; eatenCorki[2] = board.getCellValue(target);
            board.makeEmpty(target);
        }
        lastMove = landingCell;
        board.moveCorki(startCell, landingCell);
        if (board.shouldKingify(landingCell)) { // crown the new king, if there is one
            wasKingified = true;
            board.makeKing(landingCell);
        }
        return new boolean[] {wasKingified, didEat};
    }

    public boolean canMove(int[] fromCell, int[] toCell)  {
        if(!isOwnCorki(fromCell)) {
            System.out.println("NOT YOUR CORKI, CAN'T MOVE");
            return false;
        }
        if((!Player.isEgregna) && board.isKing(fromCell)) {
            return board.isEmptyInBetween(fromCell, toCell);
        } else {
            if(!board.areDiagonals(fromCell, toCell, 1)) {
                return false;
            }
            return board.isEmptyLegalCell(toCell) &&
                    (board.isKing(fromCell) ||
                            board.isForwardToTeter(fromCell, toCell));
        }
    }


    public boolean canEat(int[] fromCell, int[] landingCell){
        if(!isOwnCorki(fromCell)) {
            System.out.println("Menesha: "+Arrays.toString(fromCell));
            System.out.println("NOT YOUR CORKI, CAN'T EAT");
            return false;
        }
        if((!Player.isEgregna) && board.isKing(fromCell)) {
            if(!board.areDiagonals(fromCell, landingCell))
                return false;
            else {
                int[] prey = board.getTankegnaKingsPrey(fromCell, landingCell);
                return board.isEmptyLegalCell(landingCell) && prey != null;
            }
        } else {
            if(!board.areDiagonals(fromCell, landingCell, 2))
                return false;
            else {
                int[] nextCell = board.getMiddleCell(fromCell, landingCell);
                return canEatHelper(fromCell, nextCell, landingCell);
            }
        }

    }

    public boolean canEatHelper(int[] fromCell, int[] nextCell, int[] landingCell)  {
        if (board.areTeamMates(fromCell, nextCell) ||
                !board.areDiagonals(fromCell, nextCell, 1)) {
            return false;
        } else {
            boolean landingCellOk = landingCell != null &&
                    board.isEmptyLegalCell(landingCell);
            boolean isEatAllowed;
            if (Player.isEgregna) {
                isEatAllowed = (board.isKing(fromCell) && board.isOccupied(nextCell)) ||
                        (board.isTeter(nextCell) && board.isForwardToTeter(fromCell, nextCell));
            } else { // tankegna Teter
                isEatAllowed = board.isOccupied(nextCell);
            }
            return landingCellOk && isEatAllowed;
        }
    }

    public boolean isOwnCorki(int row, int col){
        return (isTop && board.isTop(row, col)) ||
                ((! isTop) && board.isBottom(row, col));
    }

    public boolean isOwnCorki(int[] cell){
        return isOwnCorki(cell[0], cell[1]);
    }

}

File: dama-python/player.py
"""
==================================================
THE Player IS A PROGRAM MODEL OF A DAMA PLAYER
==================================================
"""

NEIGHBOR_OFFSETS = ((-1, -1), (-1, 1), (1, 1), (1, -1))

class Player:

    # create player object
    def __init__(self, dama_board, is_top):
        self.board = dama_board
        self.is_top = is_top

        self.teter_eat_first = False

        self.teter_eats, self.king_eats = [], []
        self.teter_moves, self.king_moves = [], []

    # returns the number of corkis that a player have left on board
    def getNumCorkis(self):
        num_corkis = 0
        for row in range(self.board.NUM_ROWS):
            for col in range(self.board.NUM_ROWS):
                if self.is_top:
                    num_corkis += (self.is_top and self.board.isTop(row, col))
                else:
                    num_corkis += ((not self.is_top) and self.board.isBottom(row, col))
        return num_corkis

    # checks whether corki at the given cell belongs to the current player
    def isOwnCorki(self, row, col):
        return (self.is_top and self.board.isTop(row, col)) or \
               ((not self.is_top) and self.board.isBottom(row, col))

    '''
    goes through all corkis that belong to the comp-layer checking
    for possible moves and eats, and stores these moves and eats as
    instance variables
    '''
    def markAllMovesAndEats(self):
        teter_moves = []; king_moves = []; teter_eats = []; king_eats = []
        for row in range(self.board.NUM_ROWS):
            for col in range(self.board.NUM_ROWS):
                if self.isOwnCorki(row, col):
                    moves_and_eats = self.movesAndEats(row, col)
                    if len(moves_and_eats[1]) > 0:
                        eats = moves_and_eats[1]
                        if self.board.isTeter(row, col):
                            teter_eats.append(((row, col), eats))
                        else:
                            king_eats.append(((row, col), eats))
                    elif len(moves_and_eats[0]) > 0:
                        moves = moves_and_eats[0]
                        if self.board.isTeter(row, col):
                            teter_moves.append(((row, col), moves))
                        else:
                            king_moves.append(((row, col), moves))
        # set the corresponding instance variables
        self.teter_moves = teter_moves; self.king_moves = king_moves
        self.teter_eats = teter_eats; self.king_eats = king_eats

    # returns the cells a given corki@cur_cell can legally move to in one step, in Egregna Dama game.
    def movesAndEats(self, row, col = None):
        if col is None: cur_cell = row
        else: cur_cell = (row, col)
        legal_move_cells, legal_eat_cells = [], []
        for offset in NEIGHBOR_OFFSETS:
            neighbor_cell = cur_cell[0] + offset[0], cur_cell[1] + offset[1]
            if self.board.isInsideBoard(neighbor_cell):           
                if self.canMove(cur_cell, neighbor_cell):
                    legal_move_cells.append(neighbor_cell)
                else:
                    if self.canEat(cur_cell, neighbor_cell):
                        landing_cell = self.board.getNextCell(cur_cell, neighbor_cell)                      
                        legal_eat_cells.append(landing_cell)
        return legal_move_cells, legal_eat_cells

    # returns true if corki at from_cell can move to the to_cell
    def canMove(self, from_cell, to_cell):
        return self.board.isEmptyLegalCell(to_cell) and \
               (self.board.isKing(from_cell) or self.board.isForwardToTeter(from_cell, to_cell))

    # returns true if corki at from_cell can eat corki at to_cell
    def canEat(self, from_cell, next_cell):
        if self.board.areTeamMates(from_cell, next_cell):
            return False
        else:
            landing_cell = self.board.getNextCell(from_cell, next_cell)
            landing_cell_ok = self.board.isInsideBoard(landing_cell) and self.board.isEmptyLegalCell(landing_cell)
            is_eat_allowed = self.board.isKing(from_cell) or \
                              ( self.board.isTeter(from_cell) and self.board.isTeter(next_cell) and
                                self.board.isForwardToTeter(from_cell, next_cell) )
            return landing_cell_ok and is_eat_allowed

    # returns the next legal moves from cell
    def movesAndEatsCombined(self, cell):
        moves_and_eats = self.movesAndEats(cell)
        moves_and_eats[0].extend(moves_and_eats[1])
        return set(moves_and_eats[0])

    # returns cells whose corki can either eat or move
    def getCanAdvanceCells(self):
        can_advance_cells = []
        for row in range(self.board.NUM_ROWS):
            for col in range(self.board.NUM_ROWS):
                if self.isOwnCorki(row, col):
                    moves_and_eats = self.movesAndEats(row, col)
                    num_moves = len(moves_and_eats[0])
                    num_eats = len(moves_and_eats[1])
                    if (num_moves > 0) or (num_eats > 0):
                        can_advance_cells.append((row, col))
        return can_advance_cells

    # performs dama move from start_cell to landing_cell, returns is_kingified
    def playOneStep(self, start_cell, landing_cell, isAnEatMove):
        self.board.moveCorki(start_cell, landing_cell) # perform move
        if isAnEatMove: # eat, if it is an eat move
            mid_cell = self.board.getMiddleCell(start_cell, landing_cell)
            self.board.makeEmpty(mid_cell)
        is_kingified = False
        if self.board.shouldKingify(landing_cell): # kingify if necessary
            self.board.makeKing(landing_cell)
            is_kingified = True
        return is_kingified

class HumanPlayer(Player):
    
    def __init__(self, dama_board, is_top=False):
        super().__init__(dama_board, is_top=is_top)

    # performs dama move from start_cell to landing_cell,
    # returns is_kingified, isAnEatMove, (startcell, endcell)
    def playOneStep(self, start_cell, landing_cell, isAnEatMove=False):
        moves_and_eats = self.movesAndEats(start_cell[0], start_cell[1])
        isAnEatMove = len(moves_and_eats[1]) > 0
        is_kingified = super().playOneStep(start_cell, landing_cell, isAnEatMove)
        return is_kingified, isAnEatMove 
